Mentor View Form Validation

Claudio-Osorio opened the pull request:
Re-formatted the AddAssignment form for clarity.
Created a new view "SubmitForm". When the form is validated the data is sent to the new "SubmitForm" view. This can be changed later to accommodate for the necesities of the other areas.
Implemented basic form validation by ensuring the fields are not left blank. Text in the fields cannot be longer than specified.

Commit:
Form Validation

Re-formatted the AddAssignment form for clarity.
Created a new view "SubmitForm". When the form is validated the data is sent to the new "SubmitForm" view. This can be changed later to accommodate for the necesities of the other areas.
Implemented basic form validation by ensuring the fields are not left blank. Text in the fields cannot be longer than specified.

Claudio-Osorio committed:
Merge branch 'development' into MentorViewFormValidation

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,21 +13,41 @@ public IActionResult Index()
         [ValidateAntiForgeryToken]
         public ActionResult SubmitForm(Models.Assignment assignment)
         {
+
```
extra whitespace

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,21 +13,41 @@ public IActionResult Index()
+            return View("SubmitForm");
+        }
+
+
```
extra whitespace

jonnywwright commented on src/Mentat.UI/Mentat.UI.csproj:
```diff
@@ -9,8 +9,8 @@
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.6" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.6">
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.7" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.7">
```
These need to go. Maybe you had a reference to these tools at one point but we shouldn't need these

jonnywwright replied:
ohhhh nvm. I see now, these were not brought in by you. Just version updates. My mistake.

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -8,16 +8,51 @@ namespace Mentat.UI.Models
+
+
+
+
```
whitespace

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -8,16 +8,51 @@ namespace Mentat.UI.Models
+
+
+
+
```
whtiespace

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -8,16 +8,51 @@ namespace Mentat.UI.Models
 {
     public class Assignment
     {
+        // ASSIGNMENT ID
```
Let's remove all occurences of these comments

jonnywwright commented on src/Mentat.UI/Views/Mentor/AddAssignment.cshtml:
```diff
@@ -1,34 +1,78 @@
+}
+
+@section Scripts {
+    @{await Html.RenderPartialAsync("_ValidationScriptsPartial");}
```
I don't see a file being brought in called _ValidationScriptsPartial. Can you fill me in on this?

Claudio-Osorio replied:
Without this the validation would not show up as an overlay in the view. This is just a name autogenerated that calls some scripts form the library that already exist within the project.

jonnywwright replied:
I see, the file _ValidationScriptsPartial is already in the project. Thanks

jonnywwright commented on src/Mentat.UI/Views/Mentor/SubmitForm.cshtml:
```diff
@@ -0,0 +1,8 @@
+    Layout = "~/Views/Shared/_Layout.cshtml";
+}
+
+<h1>SubmitForm</h1>
```
Since its user facing put a space between words please.

jonnywwright replied:
Should this actually say something like 'you've successfully submitted your form'? Or something like that, since it displays after submitting the form. If that's the case just rename the file to something like Success.cshtml or something like that

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,21 +13,41 @@ public IActionResult Index()
 
         [HttpPost]
         [ValidateAntiForgeryToken]
         public ActionResult SubmitForm(Models.Assignment assignment)
```
Not a required change, but please try this ...
```csharp
public ActionResult SumitForm([FromForm] Assignment assignment)
{
    ... remove all the stuff before if (ModelState.IsValid)
} 
```

At a minimum please remove 'Models.' before the word assignment. This should be doable by importing the name-space. Same thing on line 17

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,21 +13,41 @@ public IActionResult Index()
+            // otherwise view is reloaded and user must retry filling the form
+            if (ModelState.IsValid)
+            {
+                return RedirectToAction("SendData", assignment);
```
This is confusing, since SendData is a private method below, just replace this line with 

return View("SubmitForm"); 

But give SubmitForm a new name like "Success" or something like that.

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,21 +13,41 @@ public IActionResult Index()
+        public IActionResult SendData(Models.Assignment assignment)
+        {
+            return View("SubmitForm");
+        }
```
Remove this method. This is paired with previous comment.

Claudio-Osorio replied:
This action has now been replaced with "Success"  for the much more desired  ./Success/ URL.

jonnywwright reviewed (changes requested):
Please address comments

Nickt1596 reviewed (dismissed):
LGTM

Commit:
Enhanced Form Validation and more

Added more contraints to the form validation to check for .c or .cpp file names
Fixed some typos
Remove comments and code used for demostration
Refactored the success view used after the form is submitted. Now the view has the correct URL and provides the correct feedback to the user.

Commit:
Merge branch 'MentorViewFormValidation' of https://github.com/dyohn/mentat into MentorViewFormValidation

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,20 +10,27 @@ public IActionResult Index()
+            return View("AddAssignment");
+        }
+
+        public ActionResult Success(Assignment assignment)
```
This method should be deleted.

jonnywwright commented on src/Mentat.UI/Controllers/MentorController.cs:
```diff
@@ -10,20 +10,27 @@ public IActionResult Index()
+            if (ModelState.IsValid)
+            {
+                // On Form Validation Success Go To This View
+                return RedirectToAction("Success");
```
This should return the View Success, not RedirectToAction.

Claudio-Osorio replied:
The success page/action  "/Success/" should be a page of summary where the mentor sees a live summary of the assignment during and after it is being graded. Take it as a loose end that will be used by whoever develops the assignment grading system. It can always be refactored later if they want to change the design

jonnywwright replied:
Ok sure

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -1,23 +1,106 @@
+        public FileExtensionCheckAttribute()
+        {
+            Assignment assignment = new Assignment();
+            if (assignment.AssignmentType != null)
```
You initiialize assignment on the prevoius line using the default constructor. Why are you if elsing assignment.AssignmentType? It either is or isn't null always. Did you mean to do something different on line 18?

Claudio-Osorio replied:
This will not affect the program at runtime. I'm supposed to pass in "assignment" but I had much trouble doing so. This part is only for a specific error message. It should be updated once I get help.

jonnywwright replied:
Got ya, thanks for clarification

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -1,23 +1,106 @@
+            }
+        }
+
+        public string FileType { set; get; }
```
Properties go at the top of the class definition

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -1,23 +1,106 @@
+                    fileName.Substring(fileName.Length - 2).ToLower() == ".c")
+                {
+                    return ValidationResult.Success;
+
```
whitespace

jonnywwright commented on src/Mentat.UI/Models/Assignment.cs:
```diff
@@ -1,23 +1,106 @@
 
 namespace Mentat.UI.Models
 {
+    public class FileExtensionCheckAttribute : ValidationAttribute
```
This class should go in its own file. It also needs a unit test. That said, time is pretty short. Instead can you make the class private and swap its placement on this page with Assignment. 

The top class definition belongs to the public class that shares its name with the file name.

jonnywwright replied:
Don't worry about this not a big deal.

jonnywwright reviewed (approved):
Don't worry about last few comments. Nice job with the custom validation attribute.

Nickt1596 reviewed (approved):
LGTM
